obspy/io/nordic/ellipse.py

```diff
@@ -0,0 +1,329 @@
+
+        # Rotate
+        R_rot = _ellipse.__ROT_CCW(self.theta)
+        rotated = np.dot(R_rot, pt1)
+        return rotated
+
+    def __absolute_viewpoint(self, pt):
+        """ Coordinates of a point after 'uncentering' ellipse
+
+        Assume that the ellipse was "centered" for calculations, now
+        put the point back in its true position
+
+        Inputs:
+            x,y:   coordinates of the viewpoint
+
+        Outputs:
+            x1,y1: new coordinates of the viewpoint
+        """
+        # Unrotate
+        R_rot = _ellipse.__ROT_CW(self.theta)
+        unrot = np.dot(R_rot, pt)
+        # Untranslate
+        pt1 = (unrot[0] + self.x, unrot[1]+self.y)
+        return pt1
+
+    def __get_tangents(self, pt=(0, 0)):
+        """ Return tangent intersections for a point and the ellipse
+
+        Equation is from http://www.nabla.hr/Z_MemoHU-029.htm
+        P = (-a**2 * m/c, b**2 / c)
+        """
+        if self.is_inside(pt):
+            print('No tangents, point is inside ellipse')
+            return [], []
+        elif self.is_on(pt):
+            print('No tangents, point is on ellipse')
+            return [], []
+
+        # for calculations, assume ellipse is centered at zero and pointing S-N
+        (x1, y1) = self.__relative_viewpoint(pt)
+        ms = np.roots([self.a**2 - y1**2, 2*y1*x1, self.b**2 - x1**2])
+        cs = -(y1 * ms) + x1
+        # Determine the tangent intersect with ellipse
+        T0 = (-self.a**2 * ms[0] / cs[0], self.b**2 / cs[0])
+        T1 = (-self.a**2 * ms[1] / cs[1], self.b**2 / cs[1])
+        T0 = (self.b**2 / cs[0],         -self.a**2 * ms[0] / cs[0])
+        T1 = (self.b**2 / cs[1],         -self.a**2 * ms[1] / cs[1])
+
+        # Rotate back to true coords
+        T0 = self.__absolute_viewpoint(T0)
+        T1 = self.__absolute_viewpoint(T1)
+        return T0, T1
+
+    def subtended_angle(self, pt=(0, 0), debug=False):
+        """ Find the angle subtended by an ellipse when viewed from x,y
+
+        pt=        (x,y) coordinates of the viewpoint
+
+        Equations are from http://www.nabla.hr/IA-EllipseAndLine2.htm
+        For a "centered" ellipse
+            y=mx+c
+            a^2*m^2 + b^2 = c^2
+            where   x,y are the viewpoint coordinates,
+                    a,b are the semi-* axis
+                    m and c are unknown
+            => (a^2 - x^2)*m^2 + 2*y*x*m + (b^2 - y^2) = 0  [Solve for m]
+            and then c  = y - mx
+        """
+        # If point is on or inside the ellipse, no need to calculate tangents
+        if self.is_inside(pt):
+            return 360.
+        elif self.is_on(pt):
+            return 180.
+        # Move point to origin
+        temp = copy.copy(self)
+        temp.x -= pt[0]
+        temp.y -= pt[1]
+        T0, T1 = temp.__get_tangents((0, 0))
+        cosang = np.dot(T0, T1)
+        sinang = np.linalg.norm(np.cross(T0, T1))
+        return np.degrees(np.arctan2(sinang, cosang))
+
+    def plot(self):
```

Comment:
Do we need these? If so, can you add tests for the plots?

Reply:
Actually, I think the ellipse class can be useful beyond Nordic, and these routines are useful for that.  But I prefered to start it out inside io/nordic before striking out into other, more accessed, routines.

How do I make a plot test?

Reply:
Thanks for all of your work @WayneCrawford  - I've been holding off replying to anything while you have been going, but thought I would respond here.  It's been awesome seeing so much activity on this!

Have a look at some of the other obspy image tests, like [this one](https://github.com/obspy/obspy/blob/master/obspy/imaging/tests/test_plot.py) for an example of how to do this.

Reply:
This could take me a few days, as I have other chores.  The plotting routines are not really meant to be stand-alone, but rather to add ellipses to another plot (with stations, contours, etc).

Reply:
Ah, okay. We should still make sure they are tested.  I'm happy to help however you want.  I'm not sure how to work on your fork though,  but if you are happy I can try and take a look at writing these tests on Monday?

Agree that Ellipse will be useful outside of Nordic, and leaving those other methods in would be good. We should make sure they are tested though. Again,  happy and keen to help! Thanks for all this!

Reply:
Well, I got started and couldn't stop, so now there are tests for the plotting routines, and the plotting routines have been updated to behave more like routines elsewhere in obspy